File: Difficulty: Medium/Spirally traversing a matrix/spirally-traversing-a-matrix.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
class Solution {
  public:
    vector<int> spirallyTraverse(vector<vector<int> > &matrix) {
        // code here
        // 01.08.24 POTD
        vector<int> ans;
        int n=matrix.size(), m=matrix[0].size(), l=0, r=m-1, t=0, b=n-1;
        
        while(l<=r && t<=b){
            // left to right
            for(int j=l; j<=r; j++){
                ans.push_back(matrix[t][j]);
            }
            t++;
            
            // top to bottom
            for(int i=t; i<=b; i++){
                ans.push_back(matrix[i][r]);
            }
            r--;
            
            // right to left
            if(t<=b){
                for(int j=r; j>=l; j--){
                    ans.push_back(matrix[b][j]);
                }
                b--;
            }
            
            // bottom to top
            if(l<=r){
                for(int i=b; i>=t; i--){
                    ans.push_back(matrix[i][l]);
                }
                l++;
            }
        }
        
        return ans;
        
        
    }
};

//{ Driver Code Starts.
int main() {
    int t;
    cin >> t;

    while (t--) {
        int r, c;
        cin >> r >> c;
        vector<vector<int>> matrix(r, vector<int>(c, 0));

        for (int i = 0; i < r; i++) {
            for (int j = 0; j < c; j++) {
                cin >> matrix[i][j];
            }
        }

        Solution ob;
        vector<int> result = ob.spirallyTraverse(matrix);
        for (int i = 0; i < result.size(); ++i)
            cout << result[i] << " ";
        cout << endl;
    }
    return 0;
}
// } Driver Code Ends

File: Difficulty: Medium/Spirally traversing a matrix/spirally-traversing-a-matrix.py
class Solution:
    # Function to return a list of integers denoting spiral traversal of matrix.
    def spirallyTraverse(self, mat):
        # code here
        # 20.12.24 POTD
        ans = []
        n, m = len(mat), len(mat[0])
        left, top, right, bottom = 0, 0, m-1, n-1

        while left<=right and top<=bottom:
            for j in range(left, right+1):
                ans.append(mat[top][j])
            top += 1
            
            for i in range(top, bottom+1):
                ans.append(mat[i][right])
            right -= 1
            
            if top<=bottom:
                for j in range(right, left-1, -1):
                    ans.append(mat[bottom][j])
                bottom -= 1
                
            if left<=right:
                for i in range(bottom, top-1, -1):
                    ans.append(mat[i][left])
                left += 1
                
        return ans


#{ 
 # Driver Code Starts
if __name__ == "__main__":
    import sys
    input = sys.stdin.read
    data = input().split()

    t = int(data[0])
    index = 1
    for _ in range(t):
        r = int(data[index])
        c = int(data[index + 1])
        index += 2
        matrix = []
        for i in range(r):
            row = list(map(int, data[index:index + c]))
            matrix.append(row)
            index += c

        solution = Solution()
        result = solution.spirallyTraverse(matrix)
        print(" ".join(map(str, result)))
        print("~")

# } Driver Code Ends